Document for Trash & BulkTrash operation (#397)
Laravel-Backpack/docs: 5.x/crud-operation-trash.md

# Trash Operation <span class="badge badge-info">PRO</span>

--

<a name="about"></a>
## About

This CRUD operation allows your admins to soft delete, restore and permanently delete entries from the table.

<a name="requirements"></a>
## Requirements

This is a <span class="badge badge-info">PRO</span> operation. It requires that you have [purchased access to `backpack/pro`](https://backpackforlaravel.com/products/pro-for-unlimited-projects).

In addition, it needs:

> - Your entity must use SoftDeletes to utilize this Operation.

Reviewer: ```suggestion
- Your entity must use Laravel's `SoftDeletes` trait, for this Operation to work
```

> This is a <span class="badge badge-info">PRO</span> operation. It requires that you have [purchased access to `backpack/pro`](https://backpackforlaravel.com/products/pro-for-unlimited-projects).
> 
> In addition, it needs:
> - Your entity must use Laravel's `SoftDeletes` trait, for this Operation to work

Reviewer: These could use some instructions on how to add SoftDeletes to a model, to save ourselves some questions from noobs. 

```suggestion
1. This is a <span class="badge badge-info">PRO</span> operation. It requires that you have [purchased access to `backpack/pro`](https://backpackforlaravel.com/products/pro-for-unlimited-projects).

2. In addition, it needs that your Model uses Laravel's `SoftDeletes` trait. To have your Model use SoftDeles, you should:
- generate a migration to add the `deleted_at` column to your table, eg. `php artisan make:migration add_soft_deletes_to_products --table=products`;
- inside that file's `Schema::table()` closure, add `$table->softDeletes();`
- run `php artisan migrate`
- add `use SoftDeletes` on the corresponding model (and import that class namespace);  
```

<a name="trash-a-single-item"></a>
## Trash a Single Item <span class="badge badge-info">PRO</span>

<a name="how-it-works"></a>
### How it Works
- **Trash**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/trash```, which points to the ```trash()``` method in your EntityCrudController.

- **Restore**:
Using AJAX, a PUT request is performed towards ```/entity-name/{id}/restore```, which points to the ```restore()``` method in your EntityCrudController.

- **Delete**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/delete-permanently```, which points to the ```deletePermanently()``` method in your EntityCrudController.

<a name="enabling"></a>
### How to Use

> To enable it, you need to ```use \Backpack\CRUD\app\Http\Controllers\Operations\TrashOperation;``` on your EntityCrudController. For example:

Reviewer: Let's not do that. Let's use the PRO namespace. We want CRUD to depend less on pro, not more. Is this the final namespace then?

```suggestion
To enable it, you need to ```use \Backpack\PRO\Http\Controllers\Operations\TrashOperation;``` on your EntityCrudController. For example:
```

> To enable it, you need to ```use \Backpack\PRO\Http\Controllers\Operations\TrashOperation;``` on your EntityCrudController. For example:

Reviewer: Actually it won't work if you try to use `Backpack\PRO`, you have to use `Backpack\Pro` (not all caps). Took me more time to figure this out than I would like to admit 😅

```suggestion
To enable it, you need to ```use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;``` on your EntityCrudController. For example:
```

```php
<?php

namespace App\Http\Controllers\Admin;

use Backpack\CRUD\app\Http\Controllers\CrudController;

class ProductCrudController extends CrudController
{

>     use \Backpack\CRUD\app\Http\Controllers\Operations\TrashOperation;

Reviewer: ```suggestion
    use \Backpack\PRO\Http\Controllers\Operations\TrashOperation;
```

>     use \Backpack\PRO\Http\Controllers\Operations\TrashOperation;

Reviewer: ```suggestion
    use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;
```

}
```

> 

Reviewer 2: This operation like other operations expose some access permissions that you can configure. 

When used, `TrashOperation` exposes `trash`, `restore` and `delete_permanently` access for each corresponding operation. 
Likewise, `BulkTrashOperation` has `bulkTrash`, `bulkRestore` and `bulkDeletePermanently`. 

You can for example, revoke access to some operations depending on your user roles:
```php
// if user is not superadmin, don't allow permanently delete
public function setupTrashOperation() 
{
    if(! backpack_user()->hasRole('superadmin') {
         $this->crud->denyAccess('delete_permanently');
    }
}
```

You can also disable the default trash filter:
```php
public function setupTrashOperation() 
{
    $this->crud->setOperationSetting('withTrashFilter', false);
}
```
Disabling the filter also show the trashed items in your List view and configure the buttons according to the entry delete state. 
By default the `Delete Permanently` button is only shown in trashed items, if you want to allow your admins to permanently delete without sending first to trash you can achieve that by defining in your operation setup:

```php
// in the setupTrashOperation method
CRUD::setOperationSetting('allowDeleteWithoutTrash', true);
```

> 

Reviewer 2: ```suggestion
This operation like other operations expose some access permissions that you can configure. 

When used, `TrashOperation` exposes `trash`, `restore` and `delete_permanently` access for each corresponding operation. 
Likewise, `BulkTrashOperation` has `bulkTrash`, `bulkRestore` and `bulkDeletePermanently`. 

You can for example, revoke access to some operations depending on your user roles:
```php
// if user is not superadmin, don't allow permanently delete
public function setupTrashOperation() 
{
    if(! backpack_user()->hasRole('superadmin') {
         $this->crud->denyAccess('delete_permanently');
    }
}
```

You can also disable the default trash filter:
```php
public function setupTrashOperation() 
{
    $this->crud->setOperationSetting('withTrashFilter', false);
}
```
Disabling the filter also show the trashed items in your List view and configure the buttons according to the entry delete state. 
By default the `Delete Permanently` button is only shown in trashed items, if you want to allow your admins to permanently delete without sending first to trash you can achieve that by defining in your operation setup:

```php
// in the setupTrashOperation method
CRUD::setOperationSetting('allowDeleteWithoutTrash', true);
```
```

This will make a Trash button and Trashed filter show up in the list view, and will enable the routes and functionality needed for the operation.

> 

Reviewer 2: ```suggestion
add the following text
```

This operation like other operations expose some access permissions that you can configure. 

When used, `TrashOperation` exposes `trash`, `restore` and `delete_permanently` access for each corresponding operation. 
Likewise, `BulkTrashOperation` has `bulkTrash`, `bulkRestore` and `bulkDeletePermanently`. 

You can for example, revoke access to some operations depending on your user roles:
```php
// if user is not superadmin, don't allow permanently delete
public function setupTrashOperation() 
{
    if(! backpack_user()->hasRole('superadmin') {
         $this->crud->denyAccess('delete_permanently');
    }
}
```

You can also disable the default trash filter:
```php
public function setupTrashOperation() 
{
    $this->crud->setOperationSetting('withTrashFilter', false);
}
```
Disabling the filter also show the trashed items in your List view and configure the buttons according to the entry delete state. 
By default the `Delete Permanently` button is only shown in trashed items, if you want to allow your admins to permanently delete without sending first to trash you can achieve that by defining in your operation setup:

```php
// in the setupTrashOperation method
CRUD::setOperationSetting('allowDeleteWithoutTrash', true);
```

Reviewer: Good point. I've polished your texts, added headings and merged them. Thanks!

<a name="how-to-overwrite"></a>
### How to Overwrite

In case you need to change how this operation works, just create ```trash()```, ```restore()```,```deletePermanently()``` methods in your EntityCrudController:

```php

> use \Backpack\CRUD\app\Http\Controllers\Operations\TrashOperation { trash as traitTrash; }

Reviewer: ```suggestion
use \Backpack\Pro\Http\Controllers\Operations\TrashOperation { trash as traitTrash; }
```

public function deletePermanently($id)
{
    $this->crud->hasAccessOrFail('delete_permanently');

    // your custom code here
}

public function trash($id)
{
    $this->crud->hasAccessOrFail('trash');

    // your custom code here
}

public function restore($id)
{
    $this->crud->hasAccessOrFail('restore');

    // your custom code here
}
```

> You can also overwrite the buttons by creating a file with the same name inside your ```resources/views/vendor/backpack/crud/buttons/```. You can easily publish the buttons there to make changes using:

Reviewer: ```suggestion
You can also override the buttons by creating a file with the same name inside your ```resources/views/vendor/backpack/crud/buttons/```. You can easily publish the buttons there to make changes using:
```

Author: ![Screenshot 2023-01-07 at 1 26 26 PM](https://user-images.githubusercontent.com/8214221/211140345-40af05c2-44f6-4114-9192-3f6ad0598c41.png)

@tabacitu We have used 'overwrite' in other docs. What should we do?

Reviewer: Damn it! Well... they're not technically overwriting anything... `override` is the correct one. So let's use `override` here, and make a mental note when we go through the v6 docs in 2 months to replace `overwrite` with `override` in all other places too.

```zsh
php artisan backpack:button --from=trash

php artisan backpack:button --from=restore

php artisan backpack:button --from=delete_permanently

> ```

Reviewer: Lovely. Didn't even think of that, but yes, people might want to override these.

<a name="trash-multiple-items-bulk-trash"></a>
## Trash Multiple Items (Bulk Trash) <span class="badge badge-info">PRO</span>

In addition to the button for each entry, <span class="badge badge-info">PRO</span> developers can show checkboxes next to each element, to allow their admin to trash, restore & delete multiple entries at once.

<a name="how-it-works"></a>
### How it Works

### How it Works
- **Trash**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/bulk-trash```, which points to the ```bulkTrash()``` method in your EntityCrudController.

- **Restore**:
Using AJAX, a PUT request is performed towards ```/entity-name/{id}/bulk-restore```, which points to the ```bulkRestore()``` method in your EntityCrudController.

- **Delete**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/bulk-delete```, which points to the ```bulkDelete()``` method in your EntityCrudController.

<a name="enabling"></a>
### How to Use

> You need to ```use \Backpack\CRUD\app\Http\Controllers\Operations\BulkTrashOperation;``` on your EntityCrudController.

Reviewer: ```suggestion
You need to ```use \Backpack\PRO\Http\Controllers\Operations\BulkTrashOperation;``` on your EntityCrudController.
```

> You need to ```use \Backpack\PRO\Http\Controllers\Operations\BulkTrashOperation;``` on your EntityCrudController.

Reviewer: ```suggestion
You need to ```use \Backpack\Pro\Http\Controllers\Operations\BulkTrashOperation;``` on your EntityCrudController.
```

<a name="how-to-overwrite"></a>
### How to Overwrite

In case you need to change how this operation works, just create a ```bulkTrash()``` method in your EntityCrudController:

```php

> use \Backpack\PRO\Http\Controllers\Operations\BulkTrashOperation { bulkTrash as traitBulkTrash; }

Reviewer: ```suggestion
use \Backpack\Pro\Http\Controllers\Operations\BulkTrashOperation { bulkTrash as traitBulkTrash; }
```

> use \Backpack\CRUD\app\Http\Controllers\Operations\BulkTrashOperation { bulkTrash as traitBulkTrash; }

Reviewer: ```suggestion
use \Backpack\PRO\Http\Controllers\Operations\BulkTrashOperation { bulkTrash as traitBulkTrash; }
```

public function bulkTrash($id)
{
    $this->crud->hasAccessOrFail('bulkTrash');

    // your custom code here
}

public function bulkRestore($id)
{
    $this->crud->hasAccessOrFail('bulkRestore');

    // your custom code here
}

public function bulkDelete($id)
{
    $this->crud->hasAccessOrFail('bulkDelete');

    // your custom code here
}
```

> You can also overwrite the buttons by creating a file with the same name inside your ```resources/views/vendor/backpack/crud/buttons/```. You can easily publish the buttons there to make changes using:

Reviewer: ```suggestion
You can also override the buttons by creating a file with the same name inside your ```resources/views/vendor/backpack/crud/buttons/```. You can easily publish the buttons there to make changes using:
```

- generate a migration to add the `deleted_at` column to your table, eg. `php artisan make:migration add_soft_deletes_to_products --table=products`;
- inside that file's `Schema::table()` closure, add `$table->softDeletes();`
- run `php artisan migrate`
- add `use SoftDeletes` on the corresponding model (and import that class namespace);  

<a name="trash-a-single-item"></a>
## Trash a Single Item <span class="badge badge-info">PRO</span>

<a name="how-it-works"></a>
### How it Works
- **Trash**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/trash```, which points to the ```trash()``` method in your EntityCrudController.

- **Restore**:
Using AJAX, a PUT request is performed towards ```/entity-name/{id}/restore```, which points to the ```restore()``` method in your EntityCrudController.

- **Delete**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/delete-permanently```, which points to the ```deletePermanently()``` method in your EntityCrudController.

<a name="enabling"></a>
### How to Use

**Step 1.** If your EntityCrudController uses the `DeleteOperation`, remove it. `TrashOperation` is a complete alternative to Delete.

> **Step 2.** You need to ```use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;``` inside your EntityCrudController. For example:

Reviewer: ```suggestion
**Step 2.** You need to ```use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;``` inside your EntityCrudController. Ideally the TrashOperation should be the last one that gets used. For example:
```

2. In addition, it needs that your Model uses Laravel's `SoftDeletes` trait. To have your Model use SoftDeles, you should:
- generate a migration to add the `deleted_at` column to your table, eg. `php artisan make:migration add_soft_deletes_to_products --table=products`;
- inside that file's `Schema::table()` closure, add `$table->softDeletes();`
- run `php artisan migrate`
- add `use SoftDeletes` on the corresponding model (and import that class namespace);  

<a name="trash-a-single-item"></a>
## Trash a Single Item <span class="badge badge-info">PRO</span>

<a name="how-it-works"></a>
### How it Works
- **Trash**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/trash```, which points to the ```trash()``` method in your EntityCrudController.

- **Restore**:
Using AJAX, a PUT request is performed towards ```/entity-name/{id}/restore```, which points to the ```restore()``` method in your EntityCrudController.

- **Delete**:
Using AJAX, a DELETE request is performed towards ```/entity-name/{id}/delete-permanently```, which points to the ```deletePermanently()``` method in your EntityCrudController.

<a name="enabling"></a>
### How to Use

> To enable it, you need to ```use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;``` on your EntityCrudController. For example:
> 
> ```php
> <?php
> 
> namespace App\Http\Controllers\Admin;
> 
> use Backpack\CRUD\app\Http\Controllers\CrudController;
> 
> class ProductCrudController extends CrudController
> {
>     use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;
> }
> ```
> 
> This will make a Trash button and Trashed filter show up in the list view, and will enable the routes and functionality needed for the operation.

Reviewer: ```suggestion
**Step 1.** If your EntityCrudController uses the `DeleteOperation`, remove it. `TrashOperation` is a complete alternative to Delete.

**Step 2.** You need to ```use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;``` inside your EntityCrudController. For example:

```php
class ProductCrudController extends CrudController
{
    // ... other operations
    use \Backpack\Pro\Http\Controllers\Operations\TrashOperation;
}```
This will make a Trash button and Trashed filter show up in the list view, and will enable the routes and functionality needed for the operation. If you're getting a "Trait not found" exception, make sure in the namespace you have typed `Backpack\Pro`, not `Backpack\PRO`. 
```
